content/posts/2018-12-13-designing-errors.md

```diff
@@ -0,0 +1,256 @@
+is very careful to not even mention exceptions and uses the Error term instead.
+`throws` and `throw` keywords are just a syntax sugar for working with additional return
+value. `throw` writes an error to a register and `try` reads it.
+Essentially it is like returning a `Pair` or an `Either`.
+
+> :book: Technical details are explained in
+> [the excellent article](https://www.mikeash.com/pyblog/friday-qa-2017-08-25-swift-error-handling-implementation.html)
+> by Mike Ash.
+
+At the same time,
+[there is an accepted proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0235-add-result.md)
+to introduce the `Result` type to the standard library as an alternative
+to `throw`-driven workflows. It can be seen as a more explicit, manual
+approach to propagating errors to callers. `Result` eliminates
+the function result vs. error duality via combining both of them.
+
+```swift
+enum Result<Value, Error> {
+    case value(Value)
+    case error(Error)
+}
+```
+
+# Go
+
+Fortunately or not Go doesn’t have either Java-like exceptions or
+Swift-like syntax sugar. And it is this way [by design](https://golang.org/doc/faq#exceptions).
+Explicit return values are used instead.
+
+```go
+func open(filename: String) (*File, error)
+```
+```go
+file, err := open("file.txt")
+
+if err != nil {
+    println("YARRR!")
+}
+```
+Basically, each function which might result in an error returns a pair
+of a value and an error. That’s it! Looks like a C-style error
+handling, but at least it is explicit and type-safe. The style
+is verbose, but it works surprisingly good due to the universal application.
+
+> :book: There is [a great article](https://evilmartians.com/chronicles/errors-in-go-from-denial-to-acceptance)
+> on coping with Go error handling by Sergey Alexandrovich.
+
+There is a `panic` function which might look like a Java
+unchecked exception. Using enough hacks makes it possible to catch `panic` errors
+but it is considered non-idiomatic.
+`panic` is the last call for help when things got really, really bad.
+
+```go
+panic("on the streets of London")
+```
+
+# Rust
+
+Rust takes what Go has and moves it to the next level, defining
+two error categories
+[right in the documentation](https://doc.rust-lang.org/book/ch09-00-error-handling.html):
+
+> Rust groups errors into two major categories: recoverable and unrecoverable errors.
+> For a recoverable error, such as a file not found error, it’s reasonable to report
+> the problem to the user and retry the operation.
+> Unrecoverable errors are always symptoms of bugs,
+> like trying to access a location beyond the end of an array.
+
+That’s what I like about Rust — the straightforward declaration of principles.
+
+* There are no exceptions — neither Java-like or Swift-like.
+* There is a `panic!` macros — but there are no ways to recover.
+* There is a `Result` type with a number of helper functions on top.
```

Comment:
Same as Go, it’s rather verbose to work with, this will hopefully be addressed by 2021 
https://www.ncameron.org/blog/rust-in-2022/